core/trino-main/src/main/java/io/trino/server/remotetask/RemoteTaskCleaner.java

```diff
@@ -0,0 +1,116 @@
+
+    @GuardedBy("this")
+    private boolean taskStatusFetcherStopped;
+
+    @GuardedBy("this")
+    private boolean taskInfoFetcherStopped;
+
+    @GuardedBy("this")
+    private boolean dynamidFilterFetcherStopped;
+
+    @GuardedBy("this")
+    private TaskState taskState;
+
+    public RemoteTaskCleaner(TaskId taskId, URI taskUri, HttpClient httpClient, ScheduledExecutorService executor, Supplier<SpanBuilder> spanBuilderFactory)
+    {
+        this.taskId = requireNonNull(taskId, "taskId is null");
+        this.taskUri = requireNonNull(taskUri, "taskUri is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        this.spanBuilderFactory = requireNonNull(spanBuilderFactory, "spanBuilderFactory is null");
+    }
+
+    public synchronized void markTaskStatusFetcherStopped(TaskState taskState)
+    {
+        if (taskStatusFetcherStopped) {
+            return;
+        }
+        taskStatusFetcherStopped = true;
+        this.taskState = taskState;
+        cleanupIfReady();
+    }
+
+    public synchronized void markTaskInfoFetcherStopped()
+    {
+        if (taskInfoFetcherStopped) {
+            return;
+        }
+        taskInfoFetcherStopped = true;
+        cleanupIfReady();
+    }
+
+    public synchronized void markDynamidFilterFetcherStopped()
+    {
+        if (dynamidFilterFetcherStopped) {
+            return;
+        }
+        dynamidFilterFetcherStopped = true;
+        cleanupIfReady();
+    }
+
+    @GuardedBy("this")
+    private void cleanupIfReady()
+    {
+        if (taskState != TaskState.FINISHED) {
+            // we do not perform early cleanup if task did not finish successfully.
+            // other workers may still reach out for the results; and we have no control over that.
+            return;
+        }
+        if (taskStatusFetcherStopped && taskInfoFetcherStopped && dynamidFilterFetcherStopped) {
+            scheduleCleanupRequest();
+        }
+    }
+
+    private void scheduleCleanupRequest()
+    {
+        executor.schedule(
+                () -> {
+                    Request request = preparePost()
+                            .setUri(uriBuilderFrom(taskUri)
+                                    .appendPath("/cleanup")
+                                    .build())
+                            .setSpanBuilder(spanBuilderFactory.get())
+                            .build();
+
+                    StatusResponseHandler.StatusResponse response = httpClient.execute(request, StatusResponseHandler.createStatusResponseHandler());
+                    if (response.getStatusCode() != 200) {
+                        log.warn("Failed to cleanup task %s: %s", taskId, response.getStatusCode());
+                    }
+                    return null;
+                },
+                ThreadLocalRandom.current().nextLong(TASK_CLEANUP_DELAY_VARIANCE_MILLIS),
```

Comment:
Why inject this variance here? It seems like this cleanup operation is not retried and should fire predictably at the moment it's decided to be ready for cleanup

Reply:
Just to avoid herd effect if we kill query which has lots of ongoing tasks. Possibly not necessary

Reply:
dropping for now